src/Catalyst.Core.Modules.Kvm/DeltaExecutor.cs

```diff
@@ -0,0 +1,519 @@
+            }
+
+            if (entry.IsValidDeploymentEntry && releaseSpec.IsEip2Enabled) result += GasCostOf.TxCreate;
+
+            return result;
+        }
+
+        [Todo(Improve.MissingFunctionality, "We need to agree on delta to state mapping details")]
+        private StateUpdate ToStateUpdate(Delta delta)
+        {
+            StateUpdate result = new StateUpdate
+            {
+                Difficulty = 1,
+                Number = 1,
+                Timestamp = (UInt256) delta.TimeStamp.Seconds,
+                GasLimit = (long) delta.GasLimit,
+                /* here we can read coinbase entries from the delta
+                   but we need to decide how to split fees and which one to pick for the KVM */
+                GasBeneficiary = Address.Zero,
+                GasUsed = 0L
+            };
+
+            return result;
+        }
+
+        private Address GetAccountAddress(ByteString publicKeyByteString)
+        {
+            if (publicKeyByteString == null || publicKeyByteString.IsEmpty) return null;
+
+            IPublicKey publicKey = _cryptoContext.GetPublicKeyFromBytes(publicKeyByteString.ToByteArray());
+            return publicKey.ToKvmAddress();
+        }
+
+        private void QuickFail(ContractEntry entry, ExecutionEnvironment env, ITxTracer txTracer)
+        {
+            // here we need to propagate back to Delta
+            // env.CurrentBlock.GasUsed += gasLimit;
+            if (txTracer.IsTracingReceipt)
+                txTracer.MarkAsFailed(env.ExecutingAccount, (long) entry.GasLimit, Bytes.Empty, "invalid");
+        }
+
+        private void Execute(Delta delta, ITxTracer txTracer, bool readOnly)
+        {
+            StateUpdate stateUpdate = ToStateUpdate(delta);
+            foreach (PublicEntry publicEntry in delta.PublicEntries)
+            {
+                ContractEntry contractEntry = new ContractEntry();
+                contractEntry.Base = publicEntry.Base;
+                contractEntry.Amount = publicEntry.Amount;
+                contractEntry.Data = ByteString.Empty;
+                contractEntry.GasLimit = 21000;
+                Execute(contractEntry, stateUpdate, txTracer, readOnly);
+            }
+
+            // revert state if any fails (take snapshot)
+            foreach (ContractEntry contractEntry in delta.ContractEntries) Execute(contractEntry, stateUpdate, txTracer, readOnly);
+
+            if (!readOnly)
+            {
+                // we should assign block rewards here (or in Ledger)
+                _stateProvider.CommitTree();
+                _storageProvider.CommitTrees();
+            }
+            else
+            {
+                _storageProvider.Reset();
+                _stateProvider.Reset();
+            }
+        }
+
+        private void TraceLogInvalidTx(ContractEntry entry, string reason)
+        {
+            if (_logger.IsEnabled(LogEventLevel.Verbose)) _logger.Verbose("Invalid enrty {entry} ({reason})", entry, reason);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/app/catalyst-network/Catalyst.Node/pullRequest?prid=4350015)